Translate lifting-state-up (#44)
reactjs/fr.reactjs.org: content/docs/lifting-state-up.md

  - "docs/flux-todo-list.html"
---

Plusieurs composants ont souvent besoin de refléter les mêmes données dynamiques. Nous recommandons de remonter l'état dans l'ancêtre commun le plus proche. Voyons comment ça marche.

> Dans cette section, nous allons créer un calculateur de température qui détermine si l'eau bout à une température donnée.

Author: Méprends-toi, [je check toujours quand je ne sais pas](https://www.toutelaconjugaison.com/conjugaison-verbe-bouillir.html) 😉

}
```

> [**Essayez sur CodePen**](https://codepen.io/gaearon/pen/ZXeOBm?editors=0010)

Author: C'est ce que j'avais mis initialement, ça a été corrigé sur une première review. Ca me va de laissre à l'infinitif 😄

}
```

> On peut à présent changer le composant `Calculator` pour afficher deux entrées de température :

Author: Je pensais garder "on" pour la légèreté de la phrase. Pour le reste je suis d'accord.

}
```

[**Essayer sur CodePen**](https://codepen.io/gaearon/pen/jGBryx?editors=0010)

> Nous avons maintenant deux entrées, mais quand vous entrez la température dans un des deux, l'autre ne se met pas à jour. Nous avons besoin de les garder synchronisés.

Reviewer: les garder synchronisés - > les garder synchronisées
ce sont les temperatures qui sont synchro

  render() {
    const temperature = this.state.temperature;
    // ...
```

Cependant, nous voulons que les deux entrées soient synchronisées. Quand on met à jour l'entrée en Celsius, celle en Fahrenheit doit refléter la température convertie, et vice-versa.

Avec React, partager l'état est possible en le déplaçant dans le plus proche ancêtre commun. On appelle ça « remonter l'état ». Nous allons supprimer l'état local de `TemperatureInput` et le déplacer dans le composant `Calculator`.

> Si le composant `Calculator` gère l'état, il devient la « source de vérité » pour la température des deux entrées. Il peut leur donner leur valeur afin qu'ils soient synchronisés. Comme le prop des deux composants `TemperatureInput` viennent du même composant parent `Calculator`, les deux entrées seront toujours synchronisées.

Reviewer: j'ai un doute sur `qu'ils soient synchronisés`, il ne faudrait pas que ce soit qu'elles soient synchronisées

Voyons comment ça marche étape par étape.

> D'abord, on remplace `this.state.temperature` par `this.props.temperature` dans le composant `TemperatureInput`. Maintenant, imaginons que `this.props.temperature` existe déjà, bien qu'on va devoir le passer depuis `Calculator` plus tard :

Author: Idem ici, j'ai préféré "on" pour la légèreté par rapport à "nous".

```js{3}
  render() {
    // Avant : const temperature = this.state.temperature;
    const temperature = this.props.temperature;
    // ...
```

On sait que [les props sont en lecture seule](/docs/components-and-props.html#props-are-read-only). Quand la `temperature` était dans l'état local, le composant `TemperatureInput` pouvait simplement appeler `this.setState()` pour la changer. Cependant, maintenant que `temperature` vient du parent par un prop, le composant `TemperatureInput` n'a pas de contrôle dessus.

> Avec React, on gère généralement ça en rendant le composant « contrôlé ». Comme un élément DOM `<input>` qui accepte des props `value` et `onChange`, notre `TemperatureInput` accepte `temperature` et `onTemperatureChange` dans les props depuis son parent, `Calculator`.

Reviewer: depuis son parent, `Calculator`. -> je ne pense pas que la virgule apres parent soit necessaire

Maintenant, quand le composant `TemperatureInput` veut mettre à jour la température, il appelle `this.props.onTemperatureChange`.

```js{3}
  handleChange(e) {

>     // Avant: this.setState({temperature: e.target.value});

Reviewer: il manque un espace insecable

    this.props.onTemperatureChange(e.target.value);
    // ...
```

>Note :
>
>Il n'y a pas de sens particulier aux noms des props `temperature` et `onTemperatureChange`. On aurait pu les appeler comme on voulait, comme `value` et `onChange`, qui est une convention de nommage.

> Le prop `onTemperatureChange` sera fourni dans les props par le composant parent `Calculator`, comme la prop `temperature`. Il s'occupera du changement en modifiant son état local, provocant un nouveau rendu des deux entrées avec les nouvelles valeurs. Nous allons nous pencher sur l'implémentation du nouveau composant `Calculator` très bientôt.

Reviewer: comme la prop `temperature`. -> ici tu utiliser prop au feminin. sauf qu avant tu t'en sers au masculin
j ai pas verifier mais il faudrait voir comment tout le monde le traduit et normer ca

de mon point de vue ca devrait etre feminin partout, etant derive de propriete

> Avant de regarder les changements du composant `Calculator`, récapitulons nos changements au composant `TemperatureInput`. Nous en avons supprimé l'état local, et au lieu de lire `this.state.temperature`, on lit `this.props.temperature`. Au lieu d'appeler `this.setState()` quand on veut faire un changement, on appelle `this.props.onTemperatureChange()`, qui est fourni par le composant `Calculator` :

Author: Je ne suis pas trop d'accord pour la suggestion "récapitulons nos changements du composant" : on fait des change *à quelque chose*, pas des changements *de quelque chose*, non ?

}
```

[**Essayer sur CodePen**](https://codepen.io/gaearon/pen/WZpxpz?editors=0010)

> Maintenant, quelle que soit l'entrée que vous modifiez, `this.state.temperature` et `this.state.scale` dans le composant `Calculator` seront mis à jour. L'une des deux entrées reçoit la valeur telle quelle, et l'autre est toujours recalculée à partir de l'état.

Reviewer: dans le composant `Calculator` seront mis à jour -> seront mis à jour dans le composant `Calculator` 
inverse l'ordre de la phrase

* Dans ces méthodes, le composant `Calculator` demande à React de se rafraîchir en appelant `this.setState()` avec la nouvelle valeur de l'entrée et l'unité de l'entrée modifiée.
* React appelle la méthode `render` du composant `Calculator` afin de savoir comment afficher son interface utilisateur. La valeur des deux entrées est recalculée en fonction de la température et de l'unité actuelles. La conversion de température est faite ici.
* React appelle la méthode `render` des deux composants `TemperatureInput` avec leurs nouvelles props spécifiées par le `Calculator`. React sait alors comment afficher leur interface utilisateur.
* React appelle la méthode `render` du composant `BoilingVerdict`, en lui passant la température en Celsius dans les props.
* React DOM met à jour le DOM avec le résultat correspondant à la valeur d'entrée souhaitée. L'entrée que nous venons de modifier reçoit la valeur actuelle, et l'autre entrée est mis à jour avec la température convertie.

Chaque mise à jour suit ces étapes pour que les entrées restent synchronisées.

> ## Leçons apprises {#lessons-learned}

Author: 😇

> Il ne doit y avoir qu'une seule « source de vérité » pour toute donnée qui change dans une application React. En générale, on donne un état aux composants qui en ont besoin pour s'afficher. Ensuite, si d'autres composants en ont également besoin, vous pouvez remonter l'état (*lift it up* NdT) à l'ancêtre commun le plus proche. Au lieu d'essayer de synchroniser des états entre différents composants, vous devriez vous baser sur des données qui se propagent [du haut vers le bas](/docs/state-and-lifecycle.html#the-data-flows-down).

Reviewer 2: ```suggestion
Il ne doit y avoir qu'une seule « source de vérité » pour toute donnée qui change dans une application React. En général, l’état est d’abord ajouté au composant qui en a besoin pour s'afficher. Ensuite, si d'autres composants en ont également besoin, vous pouvez faire remonter l'état dans l'ancêtre commun le plus proche. Au lieu d'essayer de synchroniser les états de différents composants, vous devriez vous baser sur des données qui se propagent [du haut vers le bas](/docs/state-and-lifecycle.html#the-data-flows-down).
```

Franchement pas mal 👏, juste quelques retouches

> Remonter l'état implique d'écrire plus de code générique (*boilerplate code*, NdT) par opposition à une liaison bidirectionnel, mais ça demande moins de travail pour trouver et isoler les bugs. Puisque tout état « vit » dans un composant et que seul ce composant peut le changer, les possibilités de bugs sont grandement réduites. De plus, vous pouvez implémenter n'importe quelle logique supplémentaire pour rejeter ou transformer la saisie des utilisateurs.

Reviewer 2: ```suggestion
Faire remonter l'état implique d'écrire plus de code générique *(boilerplate code, NdT)* qu’avec une liaison de données bidirectionnelle, mais le jeu en vaut la chandelle, car ça demande moins de travail pour trouver et isoler les bugs. Puisque tout état « vit » dans un composant et que seul ce composant peut le changer, la surface d’impact des bugs est grandement réduite. Qui plus est, vous pouvez implémenter n'importe quelle logique personnalisée pour rejeter ou transformer les saisies des utilisateurs.
```

Franchement pas mal 👏, juste quelques retouches

> Si quelque chose peut être dérivé des props ou de l'état, cette chose ne devrait probablement pas être dans l'état. Par exemple, plutôt que de stocker à la fois `celsiusValue` et `fahrenheitValue`, on stock uniquement la dernière `temperature` modifiée et son unité `scale`. La valeur de l'autre entrée peut toujours être calculée dans la méthode `render()` à partir de la valeur de l'autre entrée. Ça nous permet de nettoyer ou arrondir la valeur des autres champs sans perdre de précision sur la valeur entrée par l'utilisateur.

Reviewer 2: ```suggestion
Si quelque chose peut être dérivé des props ou de l'état, cette chose ne devrait probablement pas figurer dans l'état. Par exemple, plutôt que de stocker à la fois `celsiusValue` et `fahrenheitValue`, on stocke uniquement la dernière `temperature` modifiée et son unité `scale`. La valeur de l'autre champ peut toujours être calculée dans la méthode `render()` à partir de la valeur de ces données. Ça nous permet de vider ou d’arrondir la valeur de l’autre champ sans perdre la valeur saisie par l'utilisateur.
```

Encore pas mal du tout, mais les retouches habituelles…

> Quand vous voyez quelque chose qui ne va pas dans l'interface utilisateur, vous pouvez utiliser les [outils de développement React](https://github.com/facebook/react-devtools) pour inspecter les props et vous déplacer dans l'arborescence des composants jusqu'à trouver le composant responsable de la mise à jour de l'état. Ça vous permet de remonter à la source des bugs :

Reviewer 2: ```suggestion
Quand vous voyez quelque chose qui ne va pas dans l'UI, vous pouvez utiliser les [outils de développement React](https://github.com/facebook/react-devtools) pour examiner les props et remonter dans l'arborescence des composants jusqu'à trouver le composant responsable de la mise à jour de l'état. Ça vous permet de remonter à la source des bugs :
```
